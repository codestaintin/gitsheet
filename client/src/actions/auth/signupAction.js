import axios from 'axios';
import toastr from 'toastr';
import actionTypes from '../actionTypes';
import { decodeToken } from '../../../../shared/utils/helpers';

const signUpSuccess = (bool) => ({
  type: actionTypes.SIGNUP_SUCCESSFUL,
  bool
});

const signUpFailure = (bool) => ({
  type: actionTypes.SIGNUP_UNSUCCESSFUL,
  bool
});

const signUp = (user) => dispatch => {
  return axios.post('/api/v1/users/signup', user)
    .then((res) => {
      const { token } = res.data;
      if (decodeToken(token)) {
        window.localStorage.setItem('token', token);
      }
      dispatch(signUpSuccess(true));
      toastr.success(res.data.message);
      toastr.clear();
    })
    .catch(error => {
      if (error) {
        dispatch(signUpFailure(true));
        throw (error);
      }
    });
};

export default signUp;
