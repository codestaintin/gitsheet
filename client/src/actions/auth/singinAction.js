import axios from 'axios';
import toastr from 'toastr';
import actionTypes from '../actionTypes';
import { decodeToken, logout } from '../../../../shared/utils/helpers';

const signinSuccess = (bool) => ({
  type: actionTypes.SIGNIN_SUCCESSFUL,
  bool
});

const signinFailure = (bool) => ({
  type: actionTypes.SIGNIN_UNSUCCESSFUL,
  bool
});


const signIn = (user) => dispatch => {
  return axios.post('/api/v1/users/signin', user)
    .then((res) => {
      const { token } = res.data;
      if (decodeToken(token)) {
        window.localStorage.setItem('token', token);
      }
      dispatch(signinSuccess(true));
      toastr.success(res.data.message);
      toastr.clear();
    })
    .catch(error => {
      if (error) {
        dispatch(signinFailure(true));
        throw (error);
      }
    });
};

export const logOut = () => dispatch => {
  dispatch({ type: actionTypes.LOG_OUT_SUCCESS });
  logout();
};

export default signIn;