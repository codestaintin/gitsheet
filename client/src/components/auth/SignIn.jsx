/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import signIn from '../../actions/auth/singinAction';
import loginValidator from '../../../../shared/utils/loginValidator';

/**
 * @class SignUp
 *
 * @returns {JSX}
 */
class SignIn extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: {}
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    if (window.localStorage.token) {
      this.props.history.push('/');
    }
  }

  checkValidity() {
    const { errors, isValid } = loginValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  handleSubmit(event) {
    event.preventDefault();
    if (this.checkValidity()) {
      this.setState({ errors: {} });
      this.setState({
        email: '',
        password: ''
      });
      this.props.signIn(this.state)
        .then(() => {
          this.props.history.push('/');
        });
    }
  }

  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  render() {
    const { errors } = this.state;
    return (
      <Fragment>
        <div className="container py-5">
          <div className="row">
            <div className="col-md-12">
              <h2 className="text-center text-white mb-4 custom-header">
                Welcome back Git Sheet
              </h2>
              <div className="row">
                <div className="col-md-6 mx-auto">
                  <div className="card rounded-0">
                    <div className="card-header">
                      <h3 className="mb-0 text-center">Sign In</h3>
                    </div>
                    <div className="card-body">
                      <form className="form" onSubmit={this.handleSubmit}>
                        <div className="form-group">
                          <label htmlFor="email">Email</label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.email
                                    ? !!errors.email : false
                                }
                              )
                            }
                            name="email"
                            type="email"
                            onChange={this.handleChange}
                            value={this.state.email}
                            placeholder="Email"
                          />
                          <div className="invalid-feedback">
                            The email field is required
                          </div>
                        </div>
                        <div className="form-group">
                          <label htmlFor="password">
                            Password
                          </label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.password
                                    ? !!errors.password : false
                                }
                              )
                            }
                            type="password"
                            name="password"
                            onChange={this.handleChange}
                            value={this.state.password}
                            placeholder="Password"
                          />
                          <div className="invalid-feedback">
                            The password field is required
                          </div>
                        </div>
                        <button
                          type="submit"
                          className="btn custom-btn btn-lg float-right"
                        >
                          Sign In
                        </button>
                        <p>New here? <Link to="/signup">Sign Up</Link></p>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Fragment>
    );
  }
}

SignIn.propTypes = {
  signIn: PropTypes.func,
  history: PropTypes.object.isRequired
};

// TODO: Work here
const mapStateToProps = state => ({
  user: state.signinReducer
});

// TODO: Work here
const mapDispatchToProps = dispatch => bindActionCreators(
  { signIn }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
