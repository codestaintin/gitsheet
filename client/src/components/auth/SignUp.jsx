/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import classnames from 'classnames';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import signUpAction from '../../actions/auth/signupActions';
import registerValidator from '../../../../shared/utils/registerValidator';

/**
 * @class SignUp
 *
 * @returns {JSX}
 */
export class SignUp extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      confirmPassword: '',
      errors: {}
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    if (window.localStorage.token) {
      this.props.history.push('/hello');
    }
  }

  checkValidity() {
    const { errors, isValid } = registerValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    }
    return isValid;
  }

  handleSubmit(event) {
    event.preventDefault();
    if (this.checkValidity()) {
      this.setState({ errors: {} });
      this.setState({
        firstName: '',
        lastName: '',
        email: '',
        password: '',
        confirmPassword: ''
      });
      this.props.signUp(this.state)
        .then(() => {
          this.props.history.push('/hello');
        });
    }
  }

  handleChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  render() {
    const { errors } = this.state;
    return (
      <Fragment>
        <div className="container py-5">
          <div className="row">
            <div className="col-md-12">
              <h2 className="text-center text-white mb-4 custom-header">
                Welcome to Git Sheet
              </h2>
              <div className="row">
                <div className="col-md-6 mx-auto">
                  <div className="card rounded-0">
                    <div className="card-header">
                      <h3 className="mb-0 text-center">Sign Up</h3>
                    </div>
                    <div className="card-body">
                      <form className="form" onSubmit={this.handleSubmit}>
                        <div className="form-group">
                          <label htmlFor="firstname">First Name</label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.firstName
                                    ? !!errors.firstName : false
                                }
                              )
                            }
                            name="firstName"
                            onChange={this.handleChange}
                            value={this.state.firstName}
                          />
                          <div className="invalid-feedback">
                            First name field is required
                          </div>
                        </div>
                        <div className="form-group">
                          <label htmlFor="lastname">LastName</label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.lastName
                                    ? !!errors.lastName : false
                                }
                              )
                            }
                            name="lastName"
                            onChange={this.handleChange}
                            value={this.state.lastName}
                          />
                          <div className="invalid-feedback">
                            Last name field is required
                          </div>
                        </div>
                        <div className="form-group">
                          <label htmlFor="email">Email</label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.email
                                    ? !!errors.email : false
                                }
                              )
                            }
                            name="email"
                            type="email"
                            onChange={this.handleChange}
                            value={this.state.email}
                          />
                          <div className="invalid-feedback">
                            The email field is required
                          </div>
                        </div>
                        <div className="form-group">
                          <label htmlFor="password">
                            Password
                          </label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.password
                                    ? !!errors.password : false
                                }
                              )
                            }
                            type="password"
                            name="password"
                            onChange={this.handleChange}
                            value={this.state.password}
                          />
                          <div className="invalid-feedback">
                            The password field is required
                          </div>
                        </div>
                        <div className="form-group">
                          <label htmlFor="confirmPassword">
                            Confirm Password
                          </label>
                          <input
                            className={
                              classnames(
                                'form-control form-control-lg rounded-0', {
                                  'is-invalid': errors.confirmPassword
                                    ? !!errors.confirmPassword : false
                                }
                              )
                            }
                            type="password"
                            name="confirmPassword"
                            onChange={this.handleChange}
                            value={this.state.confirmPassword}
                          />
                          <div className="invalid-feedback">
                            The confirm password field is required
                          </div>
                        </div>
                        <button
                          type="submit"
                          className="btn custom-btn btn-lg float-right"
                        >
                          Sign Up
                        </button>
                        <p>Already a user? <Link to="/signin">Sign In</Link></p>
                      </form>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Fragment>
    );
  }
}

SignUp.propTypes = {
  signUp: PropTypes.func.isRequired,
  history: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  user: state.signUpReducer
});

const mapDispatchToProps = dispatch => bindActionCreators(
  { signUpAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(SignUp);
