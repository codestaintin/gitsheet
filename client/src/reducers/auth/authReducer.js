import actionTypes from '../../actions/actionTypes';

const initialState = {
  success: false
};

const authReducer = (state = initialState, action) => {
  switch (action.type) {
  case actionTypes.SIGNIN_SUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  case actionTypes.SIGNIN_UNSUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  case actionTypes.LOG_OUT_SUCCESS:
    return {
      ...state,
      success: false
    };
  case actionTypes.SIGNUP_SUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  case actionTypes.SIGNUP_UNSUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  default:
    return state;
  }
};

export default authReducer;