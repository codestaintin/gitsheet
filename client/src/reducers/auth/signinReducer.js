import actionTypes from '../../actions/actionTypes';

const initialState = {
  success: false,
  failure: false
};

const signinReducer = (state = initialState, action) => {
  switch (action.type) {
  case actionTypes.SIGNIN_SUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  case actionTypes.SIGNIN_UNSUCCESSFUL:
    return {
      ...state,
      failure: action.bool
    };
  case actionTypes.LOG_OUT_SUCCESS:
    return {
      ...state,
      success: false
    };
  default:
    return state;
  }
};

export default signinReducer;