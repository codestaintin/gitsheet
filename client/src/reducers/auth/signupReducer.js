import actionTypes from '../../actions/actionTypes';

const initialState = {
  success: false,
  failure: false
};

const signUpReducer = (state = initialState, action) => {
  switch (action.type) {
  case actionTypes.SIGNUP_SUCCESSFUL:
    return {
      ...state,
      success: action.bool
    };
  case actionTypes.SIGNUP_UNSUCCESSFUL:
    return {
      ...state,
      failure: action.bool
    };
  default:
    return state;
  }
};

export default signUpReducer;