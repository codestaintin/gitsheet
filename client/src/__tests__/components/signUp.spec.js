import React from 'react';
import { shallow } from 'enzyme';
import { SignUp } from '../../components/auth/SignUp';

describe('<SignUp>', () => {
  const props = {
    signUp: jest.fn(() => {}),
    history: {
      push: jest.fn(() => {})
    }
  };
  const tree = shallow(<SignUp {...props} />);
  it('should render SignUp component successfully', () => {
    expect(tree).toMatchSnapshot();
  });
  // TODO: Write more test suites when component is stable
});