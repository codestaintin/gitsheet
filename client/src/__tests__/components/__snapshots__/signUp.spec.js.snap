// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp> should render SignUp component successfully 1`] = `
<Fragment>
  <div
    className="container py-5"
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <h2
          className="text-center text-white mb-4 custom-header"
        >
          Welcome to Git Sheet
        </h2>
        <div
          className="row"
        >
          <div
            className="col-md-6 mx-auto"
          >
            <div
              className="card rounded-0"
            >
              <div
                className="card-header"
              >
                <h3
                  className="mb-0 text-center"
                >
                  Sign Up
                </h3>
              </div>
              <div
                className="card-body"
              >
                <form
                  className="form"
                  onSubmit={[Function]}
                >
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="firstname"
                    >
                      First Name
                    </label>
                    <input
                      className="form-control form-control-lg rounded-0"
                      name="firstName"
                      onChange={[Function]}
                      value=""
                    />
                    <div
                      className="invalid-feedback"
                    >
                      First name field is required
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="lastname"
                    >
                      LastName
                    </label>
                    <input
                      className="form-control form-control-lg rounded-0"
                      name="lastName"
                      onChange={[Function]}
                      value=""
                    />
                    <div
                      className="invalid-feedback"
                    >
                      Last name field is required
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="email"
                    >
                      Email
                    </label>
                    <input
                      className="form-control form-control-lg rounded-0"
                      name="email"
                      onChange={[Function]}
                      type="email"
                      value=""
                    />
                    <div
                      className="invalid-feedback"
                    >
                      The email field is required
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="password"
                    >
                      Password
                    </label>
                    <input
                      className="form-control form-control-lg rounded-0"
                      name="password"
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                    <div
                      className="invalid-feedback"
                    >
                      The password field is required
                    </div>
                  </div>
                  <div
                    className="form-group"
                  >
                    <label
                      htmlFor="confirmPassword"
                    >
                      Confirm Password
                    </label>
                    <input
                      className="form-control form-control-lg rounded-0"
                      name="confirmPassword"
                      onChange={[Function]}
                      type="password"
                      value=""
                    />
                    <div
                      className="invalid-feedback"
                    >
                      The confirm password field is required
                    </div>
                  </div>
                  <button
                    className="btn custom-btn btn-lg float-right"
                    type="submit"
                  >
                    Sign Up
                  </button>
                  <p>
                    Already a user? 
                    <Link
                      replace={false}
                      to="/signin"
                    >
                      Sign In
                    </Link>
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</Fragment>
`;
